pkg/proxy/proxy.go

```diff
@@ -307,28 +342,121 @@ func proxyNotFoundError(name string) error {
 	return fmt.Errorf("unrecognized proxy: %s", name)
 }
 
-// Exported API
+// must be called with mutex NOT held via p.proxyPortsTrigger
+func (p *Proxy) storeProxyPorts(reasons []string) {
```

Comment:
I wonder whether it would be worth to extract the whole store/restore logic into it's own struct (maybe even with a dedicated representation/ struct that represents the JSON (this way it wouldn't be required to export the fields that are written to the JSON).

Reply:
I'd not want to add the coordination needed between two types at this point. If we'd want to evolve the ProxyPort type into a different direction in future, we can always add a new type that has the exported fields later to be able to read in the file from the state file.

Reply:
ok